server/internal/calc/fourier.go

```diff
@@ -0,0 +1,96 @@
+package calc
+
+import (
+	"math"
+	"math/cmplx"
+)
+
+// FFT function (recursive implementation)
+func FastFourierTransform(a []complex128) []complex128 {
+	n := len(a)
+	if n <= 1 {
+		return a
+	}
+
+	// Evil LEAQ hacks to do math on values in a single instruction lmao. BX is
+	// `n`. So doing `-1(BX)` gives us n-1 in DX. Then doing TESTQ gives us a
+	// bitwise logical AND of the value of BX and DX. Then do a JNE so if TESTQ is
+	// not 0 then we panic.
+	// LEAQ    -1(BX), DX
+	// TESTQ   BX, DX
+	// JNE     panic.boi
+	// Ensure n is a power of two
+	if (n & (n - 1)) != 0 {
+		panic("Length of input must be a power of two")
+	}
+
+	// If n is in the CX register then
+	// SHRQ    $1, CX
+	// Is the same as n/2, SHRQ shifts the bits left by $1
+	even := make([]complex128, n/2)
+	odd := make([]complex128, n/2)
+	for i := 0; i < n/2; i++ {
+		// 		 MOVQ    main..autotmp_37+72(SP), DX
+		//     MOVQ    main.a+96(SP), SI
+		//     MOVQ    main.a+104(SP), CX
+		//     MOVQ    main..autotmp_36+48(SP), DI
+		//     XORL    BX, BX
+		//     JMP     main_fft_pc167
+		// main_fft_pc167:
+		//     CMPQ    BX, DI
+		//     JGE     main_fft_pc245
+		//     MOVQ    BX, R8
+		//     SHLQ    $1, BX
+		//     CMPQ    CX, BX
+		//     JLS     main_fft_pc637
+		//     MOVQ    R8, R9
+		//     SHLQ    $5, R8 // x32 because 16*2 = 32 **MATH** :crying:
+		//     MOVSD   (SI)(R8*1), X0
+		//     MOVSD   8(SI)(R8*1), X1
+		even[i] = a[i*2]
+		odd[i] = a[i*2+1]
+	}
+
+	fftEven := FastFourierTransform(even)
+	fftOdd := FastFourierTransform(odd)
+
+	result := make([]complex128, n)
+	for k := 0; k < n/2; k++ {
+		// -2 * math.Pi comes out to:
+		// MOVSD   $f64.c01921fb54442d18(SB), X1 = -6.283185307179586
+		t := complexExponential(-2*math.Pi*float64(k)/float64(n)) * fftOdd[k]
```

Comment:
So the parameters here make it so that this could actually be recalculated up to a point and then stored and simply used.

Reply:
1. Precalculate all of these values up to a certain point `N`.
2. Group the values into various sizes, like 128-bit, 256-bit and 512-bit. Ideally this way the precalculated values can simply be moved into a XMM, YMM or ZMM register directly as a single instruction. Then the current register for `fttOdd` can be multiplied by that other register.

So if the complex exponential was simple like `1, 2, 3, 4, 5, 6, 7, 8`. Then for AVX512 we'd have a fttOdd register like

`11, 12, 13, 14, 15, 16, 17, 18` or 8x 64 bit floats. each 128 bit pair is a complex 128.

Then we can do
```
VMULPD Z1, Z2, Z3
```

Where Z1 is the fft odd register; and Z2 is our complex exponential.

So the complex exponential register would look like:

```
1, 1, 2, 2, 3, 3, 4, 4
```

Because the values in the Z1 register are each 128 bits. So we need to multiply each 128 bit pair by the same complex exponential.

Reply:
I wonder if I could use `VMOVUPD` to just cram the precalculated complex exponentials into a 512 bit register. I would need to have them already layed out in pairs though like i have above with 1,1 2,2 3,3 etc.